### PR TITLE
netbox: do not convert cdata to table in self:call

### LINKED ISSUE #11624: `net.box.self:call` is much slower, than calling the function explicitly
Run the following script:

```lua
local clock = require('clock')
local log = require('log')
local fiber = require('fiber')
local net_box = require('net.box')

--
-- Configuration.
--
box.cfg{}
box.schema.space.create('test', {if_not_exists = true})
box.space.test:create_index('pk', {if_not_exists = true})

--
-- Perf test of the net.box call.
--
local ITERS = 1000000
local TUPLE = {1, 1, 'name'}

-- Net.box.self:call.
local start_t = clock.time();
for i = 1,ITERS do
    net_box.self:call('box.space.test:replace', {TUPLE})
end
local mean_nb_call = (clock.time() - start_t) / ITERS
fiber.yield()

-- Clean call.
start_t = clock.time();
for i = 1,ITERS do
    box.space.test:replace(TUPLE)
end
local mean_call = (clock.time() - start_t) / ITERS

log.warn('NET_BOX_CALL: %.8f', mean_nb_call)
log.warn('CLEAN_CALL: %.8f', mean_call)
log.warn('PERCENTAGE: %.5f', mean_nb_call / mean_call)
require('os').exit()
```

`net.box.call` is approximately 20% slower on my machine, than the explicit function call

```
2025-07-01 17:26:15.330 [259319] main/104/test_self_call.lua/test_self_call test_self_call.lua:34 W> NET_BOX_CALL: 0.00001163
2025-07-01 17:26:15.330 [259319] main/104/test_self_call.lua/test_self_call test_self_call.lua:35 W> CLEAN_CALL: 0.00000972
2025-07-01 17:26:15.330 [259319] main/104/test_self_call.lua/test_self_call test_self_call.lua:36 W> PERCENTAGE: 1.19662
```

This is a huge performace hit for `vshard`, which calls all user functions via `net.box` locally

### LINKED ISSUE #12421: net.box.self:call incorrectly handles nil values in return arguments
**Bug description**

net.box.self:call uses unpack(args), which leads to argument loss if there is nil value in the args.
This was pointed out by @olegrok 

Tarantool 3.7.0-entrypoint-110-g00826a42a7

**Steps to reproduce**

```
#!/usr/bin/env tarantool
function test_return() return 1, nil, 3 end
nb = require('net.box')
box.cfg{listen = 3301}
box.schema.user.grant('guest', 'super', nil, nil, {if_not_exists=true})
c = nb.connect(3301)
print(c:call('test_return'))
print(nb.self:call('test_return'))
```

How to run:

Just type `tarantool test.lua`.

**Actual behavior**
Script prints
1       cdata<void *>: NULL     3
1

**Expected behavior**

It should print
1       cdata<void *>: NULL     3
1       cdata<void *>: NULL     3

### LINKED ISSUE #12441: net.box.self:call ignores `box_tuple_extension` compat option
**Bug description**

When `compat.box_tuple_extension` is set to 'old', remote calls cast tuples to Lua tables, but `net.box.self:call` does not. This causes inconsistent behavior between local and remote calls.

Tarantool 3.7.0-entrypoint-111-gc87f39a790

**Steps to reproduce**

```
#!/usr/bin/env tarantool
function test_func() return box.tuple.new({1, 2, 3}) end

require('compat').box_tuple_extension = 'old'
box.cfg{listen = 3302}
box.schema.user.grant('guest', 'super', nil, nil, {if_not_exists=true})

local conn_remote_old = require('net.box').connect('localhost:3302')
local conn_self = require('net.box').self

print("Old")
print("Remote: " .. type(conn_remote_old:call('test_func')))
print("Self: " .. type(conn_self:call('test_func')))

require('compat').box_tuple_extension = 'new'

local conn_remote_new = require('net.box').connect('localhost:3302')
print("New")
print("Remote: " .. type(conn_remote_new:call('test_func')))
print("Self: " .. type(conn_self:call('test_func')))
```

How to run:

Just type `tarantool test.lua`.

**Actual behavior**

```
Old
Remote: table
Self: cdata
New
Remote: cdata
Self: cdata
```

**Expected behavior**

```
Old
Remote: table
Self: table
New
Remote: cdata
Self: cdata
```

Types should be consistent regardless of whether `self:call` or remote `call` is used

### PR DESCRIPTION
**Pull request by kamenkremen**
Previously, net.box.self:call() converted cdata into Lua tables using msgpack encode/decode. This workaround was introduced to make the local self call behaviour consistent with the remote call behaviour.

However, remote call behaviour has changed since the workaround introduction. Since commit cf484f071b67 ("box: support tuple formats in IPROTO responses"), remote calls return cdata tuples, but the local call was left behind, still converting cdata to Lua tables.

This patch drops the unconditional msgpack conversion from net.box.self:call. Local calls will now return cdata tuples, only converting them through msgpack if the compat.box_tuple_extension:is_old() option is set, matching the remote call behaviour.

Msgpack conversion also degrades perf a lot, so removing it should fix the performance drop, described in #11624.

Fixes #11624
Fixes #12421
Fixes #12441 

NO_DOC=bugfix

### CODE REVIEW

**Comment by kamenkremen** [comment]
I ran a perf test from #11624, here are the results:

2026-03-06 17:57:47.894 [52361] main/104/perftest.lua/perftest perftest.lua:34 W> NET_BOX_CALL: 0.00001582
2026-03-06 17:57:47.894 [52361] main/104/perftest.lua/perftest perftest.lua:35 W> CLEAN_CALL: 0.00001474
2026-03-06 17:57:47.894 [52361] main/104/perftest.lua/perftest perftest.lua:36 W> PERCENTAGE: 1.07305

I also ran the test on a version of this patch where the box_tuple_extension check was completely removed (the msgpack conversion was dropped entirely), diff:
```
diff --git a/src/box/lua/net_box.lua b/src/box/lua/net_box.lua
index 1bcb70f703..a23d36af42 100644
--- a/src/box/lua/net_box.lua
+++ b/src/box/lua/net_box.lua
@@ -1391,17 +1391,6 @@ local function handle_eval_result(status, ...)
         return box.error(E_PROC_LUA, (...))
     end
     local results = {...}
-    --
-    -- If box_tuple_extension is set to 'old', the remote net_box will convert
-    -- cdata to table. We do the same here for compatibility.
-    --
-    if compat.box_tuple_extension:is_old() then
-        for i = 1, select('#', ...) do
-            if type(results[i]) == 'cdata' then
-                results[i] = msgpack.decode(msgpack.encode(results[i]))
-            end
-        end
-    end
     return unpack(results)
 end
```

And here are the results:

2026-03-06 18:16:32.308 [56848] main/104/perftest.lua/perftest perftest.lua:34 W> NET_BOX_CALL: 0.00001513
2026-03-06 18:16:32.308 [56848] main/104/perftest.lua/perftest perftest.lua:35 W> CLEAN_CALL: 0.00001455
2026-03-06 18:16:32.308 [56848] main/104/perftest.lua/perftest perftest.lua:36 W> PERCENTAGE: 1.03965

File: test/box-luatest/gh_11624_net_box_self_call_return_type_test.lua [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    cg.server:exec(function()
+        local net_box = require('net.box')
+        local compat = require('compat')
+        local listen_uri = box.cfg.listen
+        local conn_self = net_box.self
+
+        compat.box_tuple_extension = 'new'
+        local conn_remote_new = net_box.connect(listen_uri)
+        local res_remote_new = conn_remote_new:call('return_tuple')
+        local res_self_new   = conn_self:call('return_tuple')
```
Comment (Serpentian) [inline test/box-luatest/gh_11624_net_box_self_call_return_type_test.lua]: Nit: triple space

File: test/box-luatest/gh_11624_net_box_self_call_return_type_test.lua [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+g.test_net_box_self_tuple_types = function(cg)
+    cg.server:exec(function()
+        local net_box = require('net.box')
+        local compat = require('compat')
+        local listen_uri = box.cfg.listen
+        local conn_self = net_box.self
+
+        compat.box_tuple_extension = 'new'
+        local conn_remote_new = net_box.connect(listen_uri)
```
Comment (Serpentian) [inline test/box-luatest/gh_11624_net_box_self_call_return_type_test.lua]: Nit: you probably don't need to create two separate connections for old and new, using one connection is enough, the return type should be changed on the fly, when compat is changed.

File: src/box/lua/net_box.lua [hunk 3]
```diff
@@ -7,6 +7,7 @@ local urilib   = require('uri')
 local internal = require('net.box.lib')
```
Comment (Serpentian) [inline src/box/lua/net_box.lua:7]: Nit: changelog is needed in that case. One changelog for two tickets is enough, they're about the same: fixing incompatibility between remote and local net.box calls

File: src/box/lua/net_box.lua
Comment (Serpentian) [inline src/box/lua/net_box.lua:1451]: Let's create the ticket and fix, while we're here. Don't forget to mention Oleg in the ticket as the one, who found it)

https://github.com/tarantool/vshard/pull/641#discussion_r2777326008


```lua
tarantool> function test_return() return 1, nil, 3 end
---
...

tarantool> nb = require('net.box')
---
...

tarantool> c = nb.connect(3301)
---
...

tarantool> c:call('test_return')
---
- 1
- null
- 3
...
tarantool> nb.self:call('test_return')
---
- 1
...
```

**Review by Serpentian — COMMENTED** [review]
Well done

File: test/box-luatest/gh_11624_net_box_self_call_return_type_test.lua (hunk 2 above)
Comment (kamenkremen) [reply test/box-luatest/gh_11624_net_box_self_call_return_type_test.lua]: I checked, it actually fails if I use only one connection

File: changelogs/unreleased/gh-11624-netbox-remote-and-local-call-consistency.md [hunk 4]
```diff
@@ -2,3 +2,4 @@
 
 * Fixed inconsistent return types between net.box.self:call and remote call
   (gh-11624).
+* Fixed argument loss if function return table with nil (gh-12421).
```
Comment (Serpentian) [inline changelogs/unreleased/gh-11624-netbox-remote-and-local-call-consistency.md]: Please, specify that this happens in the `net.box.self:call`. And let's better create the separate file, doesn't look good

File: test/box-luatest/gh_11624_net_box_self_call_return_type_test.lua (hunk 2 above)
Comment (Serpentian) [reply test/box-luatest/gh_11624_net_box_self_call_return_type_test.lua]: Indeed, remote calls become `table` only after reconnect happens, it's just another inconsistency, but nevermind, there're too many of them.

File: test/box-luatest/gh_12421_net_box_self_call_nil_unpack_test.lua [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    cg.server:drop()
+end)
+
+g.test_net_box_self_tuple_types = function(cg)
+    cg.server:exec(function()
+        local net_box = require('net.box')
+        local conn = net_box.self
+
+        local res = {conn:call('return_with_nil')}
+        t.assert_equals(res, {1, nil, 3})
```
Comment (Serpentian) [inline test/box-luatest/gh_12421_net_box_self_call_nil_unpack_test.lua]: Let's also test, that the return value of the remote call is equal to the `{1, nil, 3}`. But of course, `res_remote` and `res_self` are not equal, since remote returns box.null, and local one real nil.

Let's create ticket for that, one more incosistency. I see no sense in fixing it with cost of performance, so we'll probably just drop the `net_box.self:call` one day as Alexander proposed in the https://github.com/tarantool/tarantool/issues/11624#issuecomment-3024576287

File: src/box/lua/net_box.lua [hunk 6]
```diff
@@ -1390,9 +1391,15 @@ local function handle_eval_result(status, ...)
...
     end
     local results = {...}
-    for i = 1, select('#', ...) do
-        if type(results[i]) == 'cdata' then
-            results[i] = msgpack.decode(msgpack.encode(results[i]))
+    --
+    -- If box_tuple_extension is set to 'old', the remote net_box will convert
+    -- cdata to table. We do the same here for compatibility.
+    --
+    if compat.box_tuple_extension:is_old() then
```
Comment (Serpentian) [inline src/box/lua/net_box.lua]: I changed the script from the
https://github.com/tarantool/tarantool/issues/11624, so that the comparison is
more fair. Here it's

<details>
 <summary>Setup</summary>

```lua
-- require("jit.dump").start("ibxT", "/tmp/self_call.jdump")

local clock = require('clock')
local log = require('log')
local fiber = require('fiber')
local net_box = require('net.box')

--
-- Configuration.
--
box.cfg{}
box.schema.space.create('test', {if_not_exists = true})
box.space.test:create_index('pk', {if_not_exists = true})

--
-- Perf test of the net.box call.
--
local ITERS = 1000000
local TUPLE = {1, 1, 'name'}

function test_replace()
    return box.space.test:replace(TUPLE)
end

-- Net.box.self:call.
local start_t = clock.time();
for i = 1,ITERS do
    net_box.self:call('test_replace')
end
local mean_nb_call = (clock.time() - start_t) / ITERS
fiber.yield()

-- Clean call.
start_t = clock.time();
for i = 1,ITERS do
    test_replace()
end
local mean_call = (clock.time() - start_t) / ITERS

log.warn('NET_BOX_CALL: %.8f', mean_nb_call)
log.warn('CLEAN_CALL: %.8f', mean_call)
log.warn('PERCENTAGE: %.5f', mean_nb_call / mean_call)
require('os').exit()
```

The tests are ran on the following build (the commit differs):

```
Tarantool 3.7.0-entrypoint-110-g00826a42a7
Target: Linux-x86_64-RelWithDebInfo
Build options: cmake . -DCMAKE_INSTALL_PREFIX=/var/empty/local -DENABLE_BACKTRACE=TRUE
Compiler: GNU-12.3.0
C_FLAGS: -fexceptions -funwind-tables -fasynchronous-unwind-tables -fno-common -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/home/serpentian/Programming/tnt/tarantool=. -std=c11 -Wall -Wextra -Wno-gnu-alignof-expression -fno-gnu89-inline -Wno-cast-function-type -O2 -g -DNDEBUG -ggdb -O2
CXX_FLAGS: -fexceptions -funwind-tables -fasynchronous-unwind-tables -fno-common -msse2 -Wformat -Wformat-security -Werror=format-security -fstack-protector-strong -fPIC -fmacro-prefix-map=/home/serpentian/Programming/tnt/tarantool=. -std=c++17 -Wall -Wextra -Wno-invalid-offsetof -Wno-gnu-alignof-expression -Wno-cast-function-type -O2 -g -DNDEBUG -ggdb -O2
```

</details>

Before patchset:

```
2026-03-19 19:27:54.681 [371920] main/104/test_perf.lua/test_perf test_perf.lua:38 W> NET_BOX_CALL: 0.00001088
2026-03-19 19:27:54.681 [371920] main/104/test_perf.lua/test_perf test_perf.lua:39 W> CLEAN_CALL: 0.00000920
2026-03-19 19:27:54.681 [371920] main/104/test_perf.lua/test_perf test_perf.lua:40 W> PERCENTAGE: 1.18291
```

After the first commit:
```
2026-03-19 19:31:28.241 [384305] main/104/test_perf.lua/test_perf test_perf.lua:38 W> NET_BOX_CALL: 0.00001057
2026-03-19 19:31:28.241 [384305] main/104/test_perf.lua/test_perf test_perf.lua:39 W> CLEAN_CALL: 0.00000909
2026-03-19 19:31:28.241 [384305] main/104/test_perf.lua/test_perf test_perf.lua:40 W> PERCENTAGE: 1.16211
```

After the second one:
```
2026-03-19 19:33:04.291 [391541] main/104/test_perf.lua/test_perf test_perf.lua:38 W> NET_BOX_CALL: 0.00001086
2026-03-19 19:33:04.291 [391541] main/104/test_perf.lua/test_perf test_perf.lua:39 W> CLEAN_CALL: 0.00000929
2026-03-19 19:33:04.291 [391541] main/104/test_perf.lua/test_perf test_perf.lua:40 W> PERCENTAGE: 1.16941
```

After patch and two commits:

```diff
diff --git a/src/box/lua/net_box.lua b/src/box/lua/net_box.lua
index 3215b675fa..8c1e4c6761 100644
--- a/src/box/lua/net_box.lua
+++ b/src/box/lua/net_box.lua
@@ -1390,19 +1390,7 @@ local function handle_eval_result(status, ...)
         rollback()
         return box.error(E_PROC_LUA, (...))
     end
-    local results = {...}
-    --
-    -- If box_tuple_extension is set to 'old', the remote net_box will convert
-    -- cdata to table. We do the same here for compatibility.
-    --
-    if compat.box_tuple_extension:is_old() then
-        for i = 1, select('#', ...) do
-            if type(results[i]) == 'cdata' then
-                results[i] = msgpack.decode(msgpack.encode(results[i]))
-            end
-        end
-    end
-    return unpack(results, 1, table.maxn(results))
+    return unpack({...}, 1, table.maxn({...}))
 end

 this_module.self = {
```

```
2026-03-19 19:34:05.001 [402928] main/104/test_perf.lua/test_perf test_perf.lua:38 W> NET_BOX_CALL: 0.00001018
2026-03-19 19:34:05.001 [402928] main/104/test_perf.lua/test_perf test_perf.lua:39 W> CLEAN_CALL: 0.00000909
2026-03-19 19:34:05.001 [402928] main/104/test_perf.lua/test_perf test_perf.lua:40 W> PERCENTAGE: 1.11934
```

Comment (Serpentian) [reply src/box/lua/net_box.lua]: We probably should drop the compat completely and that code with it, or introduce just new incosistency, when that compat is `old`

**Review by Serpentian — COMMENTED** [review]
Let me think and discuss with others, what should we do with the compat: that `if` in the `handle_eval_result` majorly affects perf

File: src/box/lua/net_box.lua (hunk 6 above)
Comment (Serpentian) [reply src/box/lua/net_box.lua]: Nah, we cannot, it'll be dropped only in 4.X. Let's instead drop that code and create ticket, we intentionally create one more inconsistency, just another reason to drop that `net.box.self:call` completely, as Alexander proposes in the https://github.com/tarantool/tarantool/issues/11624#issuecomment-3024576287

Can you please, create the ticket in github about one more incompatibility, that'll appear after we merge the PR. And please, let's also create ticket about dropping the net.box.self.call in github too, mention JIRA in it and all the associated inconsistencies we know about (already fixed and not), so that we don't forget.

Comment (Serpentian) [reply src/box/lua/net_box.lua]: https://github.com/tarantool/tarantool/issues/12441

File: test/box-luatest/gh_12421_net_box_self_call_nil_unpack_test.lua [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        local conn_self = net_box.self
+
+        local conn_remote = net_box.connect(listen_uri)
+        local res_remote = {conn_remote:call('return_with_nil')}
+        local res_self = {conn_self:call('return_with_nil')}
+        conn_remote:close()
+
+        t.assert_equals(res_self, {1, nil, 3})
+        t.assert_equals(res_remote, {1, nil, 3})
+        -- There is an inconsistency between local call and remote call.
```
Comment (Serpentian) [inline test/box-luatest/gh_12421_net_box_self_call_nil_unpack_test.lua]: Let's either not test the incosistency or mention the ticket number in the comment

File: src/box/lua/net_box.lua
Comment (Serpentian) [inline src/box/lua/net_box.lua:4]: Please, update the first commit msg, and let's mention, why we didn't use `compat` and mention the newly created ticket number

**Review by Serpentian — APPROVED** [review]
Thank you a lot!

**Review by sergepetrenko — COMMENTED** [review]
Hi! Thanks for the patch!

Have you considered overloading `msgpack.encode/decode` via `compat.option.action` ?

This way the performance loss from checking an extra `if compat_option_is_new` would be fixed, I suppose.

I think it's worth trying out, at least. Even though we consider this removal of encoding/decoding good, it's a behaviour change for existing users and we should try to wrap it in compat.

I mean something like:

```lua
-- in net_box.lua

local decode_func = nil

compat.add_option{
    name = 'net_box_self_drop_conversions',
    ...
    action = function(is_new)
        if is_new then
            decode_func = function(args) return args end
        else
            decode_func = function(args)
                return msgpack.decode(msgpack.encode(args))
            end
        end
    end,
}
```

And in `handle_eval_result`:
```lua
return unpack(decode_func({...}))
```

I also think this should be a separate compat option, let's not depend on `box_tuple_extension`. It's not good to add behavior changes to existing compat options. This renders the whole idea of compat pointlesss.

**Comment by Serpentian** [comment]
> Have you considered overloading msgpack.encode/decode via compat.option.action ?

Agree, we should do that and mark the commit as though it fixes the https://github.com/tarantool/tarantool/issues/12441.

> I also think this should be a separate compat option, let's not depend on box_tuple_extension. It's not good to add behavior changes to existing compat options. This renders the whole idea of compat pointlesss.

@sergepetrenko, I cannot agree here. We have the following behavior now: with old compat the remote call returns table, local one - tables, with the `new` - remote call - cdata, local one - table. It's explicitly the bug, related to the `box_tuple_extension` compat and we should not introduce the new one. Otherwise (if new one is added), a user has to switch two compats in order to have the same behavior on local and remote calls. It seems to me that we should do that under old compat and not introduce the new one, WDYT?

File: src/box/lua/net_box.lua
Comment (Serpentian) [inline src/box/lua/net_box.lua:1]: We must figure out, how compat affects remote calls. If it makes all cdata converted, then we should introduce new comat. Otherwise, we should introduce new one

Comment (kamenkremen) [reply src/box/lua/net_box.lua:1]: It seems that `box_tuple_extension` affects only tuples. This compat option controls whether box.tuple objects are encoded as `MP_TUPLE` extension type or `MP_ARRAY`. Other types have their own extension types, so they are unaffected. For example this code
```
box.cfg{listen = 3301}
local netbox = require("net.box")
rawset(_G, 'return_error', function()
    return box.error.new{code = 100, reason = "test error"}
end)

local c = netbox:connect('localhost:3301')
local ok, res = pcall(function() return c:call('return_error') end)
print("res type:", type(res))

require('compat').box_tuple_extension = 'old'
local c2 = netbox:connect('localhost:3301')
local ok, res = pcall(function() return c2:call('return_error') end)
print("res type:", type(res))
```
will result in 
```
res type:       cdata
res type:       cdata
```

File: changelogs/unreleased/gh-11624-netbox-remote-and-local-call-consistency.md [hunk 8]
```diff
@@ -0,0 +1,4 @@
+## bugfix/netbox
+
+* Fixed inconsistent return types between net.box.self:call and remote call
+  (gh-11624).
```
Comment (AArdeev) [inline changelogs/unreleased/gh-11624-netbox-remote-and-local-call-consistency.md:4]: * Fixed inconsistent return types between `net.box.self:call()` and remote calls
  (gh-11624).

File: changelogs/unreleased/gh-12421-netbox-local-call-argument-loss-with-nil.md [hunk 9]
```diff
@@ -0,0 +1,4 @@
+## bugfix/netbox
+
+* Fixed argument loss if function return table with nil in
+  net.box.self:call(gh-12421).
```
Comment (AArdeev) [inline changelogs/unreleased/gh-12421-netbox-local-call-argument-loss-with-nil.md:4]: * Fixed argument loss when a function returns a table containing `nil` in
`net.box.self:call()` (gh-12421).

**Review by AArdeev — APPROVED** [review]

File: src/box/lua/net_box.lua [hunk 10]
```diff
@@ -7,6 +7,7 @@ local urilib   = require('uri')
 local internal = require('net.box.lib')
 local trigger  = require('internal.trigger')
 local utils    = require('internal.utils')
+local compat   = require('compat')
 
 local this_module
```
Comment (Serpentian) [inline src/box/lua/net_box.lua:13]: Let's properly split the commits:

https://github.com/tarantool/tarantool/issues/11624 and https://github.com/tarantool/tarantool/issues/12441 - first one

https://github.com/tarantool/tarantool/issues/12421 - second one
